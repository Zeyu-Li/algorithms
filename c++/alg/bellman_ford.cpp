/* Bellman-Ford */ 
#include <iostream>
#include <vector>
#include <iostream>
#include <limits.h>
#include "../structs/graph.cpp"

class Graph_bf: public Graph {
    public:
        Graph_bf(): Graph() {};
        Graph_bf(int V, int E): Graph(V, E) {};
        // stores the results the the bellman ford given source node, (no return value)
        void bell_ford(int, std::vector<int>&);
        // returns a vector of bellman ford results
        std::vector<int> bell_ford(int);
        // returns the length of the path from a source to the destination
        int bell_ford(int, int);
};

int Graph_bf::bell_ford(int start, int end) {
    std::vector<int> result;
    bell_ford(start, result);
    return result[end];
}

std::vector<int> Graph_bf::bell_ford(int start) {
    std::vector<int> result;
    bell_ford(start, result);
    return result;
}

void Graph_bf::bell_ford(int start, std::vector<int>&path) {
    // uses dynamic programming to determine shortest distance
    std::vector<int> dist(V);
    // all nodes unknown so infinite distance
    std::fill_n(dist.begin(), V, INT_MAX);
    // set current to 0
    dist[start] = 0;

    // relaxes all edges V-1 times applying dp
    for (int i = 1; i < V; i++) {
        // for every edges (this makes it O(V+E))
        for (int j = 0; j < E; j++) {
            int u = edges[j].second.first;
            int v = edges[j].second.second;
            int w = edges[j].first;
            // if edge v (destination) is connected and 
            // less than the source node + weight, update destination node 
            if (dist[u] != INT_MAX && dist[u] + w < dist[v]) dist[v] = dist[u] + w;
        }
    }

    // check for negative weighted cycles (assumed non negative weight cycles)
    // for (int j = 0; j < E; j++) { 
    //     int u = edges[j].second.first;
    //     int v = edges[j].second.second;
    //     int w = edges[j].first;
    //     if (dist[u] != INT_MAX && dist[u] + w < dist[v]) { 
    //         printf("Graph contains negative weight cycle"); 
    //         // If negative cycle is detected, error out
    //         throw -1;
    //         return;
    //     } 
    // } 

    // copy dist to path (weighted dp table)
    for(int i=0; i<dist.size(); ++i) path.push_back(dist[i]);
    return;
}
