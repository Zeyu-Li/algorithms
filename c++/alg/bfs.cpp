#include <iostream>

using namespace std;



int main(int argc, char const *argv[])
{
    cout << lcd(21,42) << endl;
    return 0;
}