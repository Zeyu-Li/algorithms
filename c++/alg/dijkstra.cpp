/* Dijkstra's Algorithm */ 
#include <vector>
#include <algorithm>
#include <limits.h> 
#include "../structs/matrix_graph.cpp"

class Graph_dijkstra: public Graph_m {
    public:
        Graph_dijkstra(): Graph_m() {};
        Graph_dijkstra(bool a): Graph_m(a) {};
        Graph_dijkstra(int a): Graph_m(a) {};
        Graph_dijkstra(bool a, int b): Graph_m(a, b) {};
        // returns a vector of the result of dijkstra's alg
        std::vector<int> dijkstra(int);
        // stores dijkstra results in second argument
        void dijkstra(int, std::vector<int>&);
        // returns shortest path given start and end
        int dijkstra(int, int);
    private:
        // gets min distance
        int min_dist(std::vector<int>&, std::vector<bool>&);
};

int Graph_dijkstra::min_dist(std::vector<int>& dist, std::vector<bool>& shortest_path) {
    int min = INT_MAX, min_index;
    // for each vertex
    for (int v = 0; v < size; v++) {
        // if is connected and smaller than current min, swap to that one
        if ((!shortest_path[v]) && dist[v] <= min) {
            min = dist[v];
            min_index = v;
        }
    }
    return min_index;
}


std::vector<int> Graph_dijkstra::dijkstra(int start) {
    std::vector<int> dist(size); // shortest path distance of each point
    std::fill_n(dist.begin(), size, INT_MAX);
    std::vector<bool> shortest_path(size);
    std::fill_n(shortest_path.begin(), size, false);

    // start of path is of distance 0
    dist[start] = 0;
    // for every vertex
    for (int i = 0; i < size - 1; i++) {
        // gets shortest path to work off of
        int u = min_dist(dist, shortest_path);
        // check the current node
        shortest_path[u] = true;

        // update dist of adjacent vertices to the shortest path if it is connect and shorter than previous entry (dp)
        for (int v = 0; v < size; v++) {
            if (!shortest_path[v] && matrix[u][v] && dist[u] != INT_MAX && dist[u] + matrix[u][v] < dist[v]) {
                dist[v] = dist[u] + matrix[u][v];
            }
        }
    }
    // return the distances of all nodes
    return dist;
}

void Graph_dijkstra::dijkstra(int start, std::vector<int>&path) {
    std::vector<int> d_path = dijkstra(start);
    // copy vectors
    for(int i=0; i<d_path.size(); ++i) path.push_back(d_path[i]);
}

int Graph_dijkstra::dijkstra(int start, int end) {
    std::vector<int> d_path = dijkstra(start);
    // return length of shortest path
    return d_path[end];
}
