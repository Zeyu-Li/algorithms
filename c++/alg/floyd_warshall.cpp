/* 
* Floyd-Warshall
* Modified from: https://www.geeksforgeeks.org/floyd-warshall-algorithm-dp-16/
*/ 
#include <vector>
#include "../structs/matrix_graph.cpp"

#define INF 99999

class Graph_fw: public Graph_m {
    public:
        Graph_fw(): Graph_m() {};
        Graph_fw(bool a): Graph_m(a) {};
        Graph_fw(int a): Graph_m(a) {};
        Graph_fw(bool a, int b): Graph_m(a, b) {};
        // from the source node will return a 
        // matrix of all optimal paths from 1 node to another
        std::vector<std::vector<int>> floyd_warshall(void);
        // stores optimal path matrix in first and only arg, no return
        void floyd_warshall(std::vector<std::vector<int>>&);
};

void Graph_fw::floyd_warshall(std::vector<std::vector<int>>& matrix) {
    matrix = floyd_warshall();
}

std::vector<std::vector<int>> Graph_fw::floyd_warshall() {
    // refactor matrix such that unconnected paths are MAX INT instead of 0
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            // cannot use max int because addition will wrap over
            if (i != j && matrix[i][j] == 0) matrix[i][j] = INF;
        }
    }

    // the dp soltution to every pair 
    std::vector<std::vector<int>> dist(size, std::vector<int>(size));
    // copy vector over
    dist = matrix;

    // minimizes for every connection for every node
    for (int k = 0; k < size; k++) {
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                if (dist[i][k] + dist[k][j] < dist[i][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }

    return dist;
}