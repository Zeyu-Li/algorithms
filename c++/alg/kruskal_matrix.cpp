/* 
* Kruskal's Algorithm for Matrix/Table Graph
* Modified from: https://www.geeksforgeeks.org/kruskals-algorithm-simple-implementation-for-adjacency-matrix/
*/ 
// Already in matrix_graph.cpp
// #include <vector>
#include <utility>
#include <algorithm>
#include <limits.h>
#include "../structs/matrix_graph.cpp"

class Graph_kruskal_m: public Graph_m {
    public:
        Graph_kruskal_m(): Graph_m() {};
        Graph_kruskal_m(bool a): Graph_m(a) {};
        Graph_kruskal_m(int a): Graph_m(a) {};
        Graph_kruskal_m(bool a, int b): Graph_m(a, b) {};
        // applied kruskal's algorithm to get MST 
        // (MST will be stored in a vector of 
        // pair(int weight, pair(int source, int destination))) 
        void kruskal(void);
        // minimum spanning tree
        std::vector<std::pair<int, std::pair<int, int>>> MST;
        void print_mst();
    private:
        int find(int);
        void union_(int, int);
        std::vector<int> parent;
};

int Graph_kruskal_m::find(int i) {
    while (parent[i] != i) {
        i = parent[i];
    }
    return i;
}

void Graph_kruskal_m::union_(int i, int j) {
    int a = find(i);
    int b = find(j);
    parent[a] = b;
}

void Graph_kruskal_m::kruskal() {
    // refactor matrix such that unconnected paths are MAX INT instead of 0
    for (int i = 0; i < size; i++) {
        for (int j = 0; j < size; j++) {
            // cannot use max int because addition will wrap over
            if (matrix[i][j] == 0) matrix[i][j] = INT_MAX;
        }
    }
    // resize vector
    parent.resize(size);
    // initialize disjoint set
    for (int i = 0; i < size; i++) parent[i] = i;

    for (int edge_count = 0; edge_count < size - 1; edge_count++) {
        int min = INT_MAX, a, b;
        a = b = -1;
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                if (find(i) != find(j) && matrix[i][j] < min) {
                    min = matrix[i][j];
                    a = i;
                    b = j;
                }
            }
        }

        // join union a and b
        union_(a, b);
        MST.push_back({min, {a, b}});
    }
}

void Graph_kruskal_m::print_mst(void) {
    std::vector<std::pair<int, std::pair<int, int>>>::iterator it;
    for(it = MST.begin();it!=MST.end();it++){
        std::cout << it->second.first << " - " << it->second.second << " (" << it->first << ")" << std::endl;
    }
}
