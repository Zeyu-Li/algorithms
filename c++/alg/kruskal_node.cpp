/* 
* Kruskal's Algorithm for Edged Based Graph
* Modified from: https://gist.github.com/MagallanesFito/791f736a0d21708794aafa11a0416201
*/ 
// Already in graph.cpp
// #include <vector>
// #include <utility>
#include <algorithm>
#include "../structs/graph.cpp"

class Graph_kruskal_n: public Graph {
    public:
        Graph_kruskal_n(): Graph() {};
        Graph_kruskal_n(int V, int E): Graph(V, E) {};
        // applied kruskal's algorithm to get MST 
        // (MST will be stored in a vector of 
        // pair(int weight, pair(int source, int destination))) 
        void kruskal(void);
        // minimum spanning tree
        std::vector<std::pair<int, std::pair<int, int>>> MST;
        void print_mst();
    private:
    // from https://gist.github.com/MagallanesFito/791f736a0d21708794aafa11a0416201#file-kruskal-cpp-L32
        struct Disjoint_set{
            int *parent,*rnk;
            int n;

            Disjoint_set(int n){
                this->n = n;
                parent = new int[n+1];
                rnk = new int[n+1];

                for(int i=0;i<=n;i++){
                    rnk[i] = 0;
                    parent[i] = i;
                }
            }
            ~Disjoint_set() {
                delete parent;
                delete rnk;
            }
            int Find(int u){
                if(u == parent[u]) return parent[u];
                else return Find(parent[u]);
            }
            // joins x and y groups
            void Union(int x,int y){
                if(x != y){
                    if(rnk[x] < rnk[y]){
                        rnk[y] += rnk[x];
                        parent[x] = y;
                    }
                    else{
                        rnk[x] += rnk[y];
                        parent[y] = x;
                    }
                }
            }
        };
};

void Graph_kruskal_n::kruskal() {
    // sort edges so to determine which edge to do first
    sort(edges.begin(),edges.end());

    // new instance of disjoinset
    Disjoint_set ds(V);

    // for all edges
    for (std::vector<std::pair<int,std::pair<int,int>>>::iterator it = edges.begin(); it != edges.end(); it++) {
        // source and destination
        int u = it->second.first;
        int v = it->second.second;

        // find u and v in disjoint set
        int set_u = ds.Find(u);
        int set_v = ds.Find(v);

        // if not in set, add to MST
        if (set_u != set_v) {
            int w = it->first;
            MST.push_back({w, {u, v}});
            
            ds.Union(set_u, set_v);
        }
    }
}

void Graph_kruskal_n::print_mst(void) {
    std::vector<std::pair<int, std::pair<int, int>>>::iterator it;
    for(it = MST.begin();it!=MST.end();it++){
        std::cout << it->second.first << " - " << it->second.second << " (" << it->first << ")" << std::endl;
    }
}
