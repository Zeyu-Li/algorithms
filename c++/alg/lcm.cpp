#include "gcd.cpp"

int lcd(int gcd, int a, int b) {
    return (int)((a*b)/gcd);
}
int lcd(int a, int b) {
    // function overload
    return (int)((a*b)/gcd(a,b));
}

// different name, same thing
int lcm(int gcd, int a, int b) {
    return (int)((a*b)/gcd);
}
int lcm(int a, int b) {
    // function overload
    return (int)((a*b)/gcd(a,b));
}
