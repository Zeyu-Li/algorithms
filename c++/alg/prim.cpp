/* 
* Prim's Algorithm
* Modified from: https://www.geeksforgeeks.org/prims-minimum-spanning-tree-mst-greedy-algo-5/
*/ 
// Already in matrix_graph.cpp
// #include <vector>
#include <utility>
#include <algorithm>
#include <limits.h>
#include "../structs/matrix_graph.cpp"

class Graph_prim: public Graph_m {
    public:
        Graph_prim(): Graph_m() {};
        Graph_prim(bool a): Graph_m(a) {};
        Graph_prim(int a): Graph_m(a) {};
        Graph_prim(bool a, int b): Graph_m(a, b) {};
        // applied prim's algorithm to get MST 
        // (MST will be stored in a vector of 
        // pair(int weight, pair(int source, int destination))) 
        void prim(void);
        // minimum spanning tree
        std::vector<std::pair<int, std::pair<int, int>>> MST;
        void print_mst();
    private:
        int min_key(std::vector<int>&, std::vector<bool>&);
        std::vector<int> parent;
};

int Graph_prim::min_key(std::vector<int>& key, std::vector<bool>& set_MST) {
    int min = INT_MAX, min_index;
    for (int v = 0; v < size; v++) {
        // if not in the mst and lower than the current min, set new min
        if (!set_MST[v] && key[v] < min) {
            min = key[v];
            min_index = v;
        }
    }
    return min_index;
}

void Graph_prim::prim() {
    // resize vector
    parent.resize(size);
    
    // key for picking minimum weight edge in cut
    std::vector<int> key(size);
    // represent the edges in the MST if counted
    std::vector<bool> set_MST(size);
    // inits keys as infinite and all vertices as not visited
    std::fill_n(key.begin(), size, INT_MAX);
    std::fill_n(set_MST.begin(), size, false);

    // starting node
    key[0] = 0;
    parent[0] = -1;

    // for every node
    for (int count = 0; count < size - 1; count ++) {
        // pick minimum key from vertices not in MST (that are connected)
        int u = min_key(key, set_MST);

        // picked vertex in mst set
        set_MST[u] = true;
        for (int v = 0; v < size; v++) {
            if (matrix[u][v] && !set_MST[v] && matrix[u][v] < key[v]) {
                parent[v] = u;
                key[v] = matrix[u][v];
            }
        }
    }

    // store in MST
    for (int i = 1; i < size; i++) {
        MST.push_back({matrix[i][parent[i]], {parent[i], i}});
    }
}

void Graph_prim::print_mst(void) {
    // std::vector<std::pair<int, std::pair<int, int>>>::iterator it;
    for(auto it = MST.begin();it!=MST.end();it++){
        std::cout << it->second.first << " - " << it->second.second << " (" << it->first << ")" << std::endl;
    }
}
