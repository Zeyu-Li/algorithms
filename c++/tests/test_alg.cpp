/* 
* Testing Algorithms
* Description: A test for all the algorithm classes
*/ 
#include <iostream>
#include <limits.h>
#include "../alg/lcm.cpp"
#include "../alg/q_sqrt.cpp"
#include "../alg/bfs.cpp"
#include "../alg/dijkstra.cpp"
#include "../alg/bellman_ford.cpp"
#include "../alg/floyd_warshall.cpp"
#include "../alg/kruskal_node.cpp"
#include "../alg/kruskal_matrix.cpp"
#include "../alg/prim.cpp"

using namespace std;

void print(int size, int arr[]) {
    for (int i = 0; i < size - 2; i++) {
        std::cout << arr[i] << " -> ";
    }
    std::cout << arr[size - 1];
}

int main() {
    /* gcd and lcm/lcd */
    int a = 21, b = 42, c = 75;
    cout << "GCD:" << endl;
    cout << gcd(a,b) << endl;
    cout << "LCM:" << endl;
    cout << lcm(a,c) << endl;

    /* quick square root */
    float num = 10;
    cout << "Quick Square Root:\n" << q_sqrt(num) << " : " << q_sqrt2(num) << endl;

    /* BFS */
    cout << "BFS:" << endl;
    int size = 8;
    Graph_bfs graph_l(true, size);
    graph_l.add_edge(0, 1);
    graph_l.add_edge(0, 3);
    graph_l.add_edge(1, 2);
    graph_l.add_edge(3, 4);
    graph_l.add_edge(3, 7);
    graph_l.add_edge(4, 5);
    graph_l.add_edge(4, 7);
    graph_l.add_edge(5, 6);
    graph_l.add_edge(6, 7);
    int source = 0, dest = 7;
    vector<int> path_bfs1;
    vector<int> path_bfs2;
    // graph_l.print();
    graph_l.bfs(source, path_bfs1);
    cout << "Length: " << graph_l.bfs_path(source, dest, path_bfs2) << endl;

    // print
    cout << "BFS order: ";
    for(int i=0; i<path_bfs1.size(); ++i) cout << path_bfs1[i] << ' ';
    cout << endl;
    cout << "Shortest path: ";
    for(int i=0; i<path_bfs2.size(); ++i) cout << path_bfs2[i] << ' ';
    cout << endl;

    /* Dijkstra */
    cout << "Dijkstra:" << endl;
    Graph_dijkstra Dijkstra(false, 9);
    Dijkstra.add_edge(0, 1, 4);
    Dijkstra.add_edge(0, 7, 8);
    Dijkstra.add_edge(1, 2, 8);
    Dijkstra.add_edge(1, 7, 11);
    Dijkstra.add_edge(2, 3, 7);
    Dijkstra.add_edge(2, 5, 4);
    Dijkstra.add_edge(2, 8, 2);
    Dijkstra.add_edge(3, 4, 9);
    Dijkstra.add_edge(3, 5, 14);
    Dijkstra.add_edge(4, 5, 10);
    Dijkstra.add_edge(5, 6, 2);
    Dijkstra.add_edge(6, 7, 1);
    Dijkstra.add_edge(6, 8, 6);
    Dijkstra.add_edge(7, 8, 7);

    // Dijkstra.print();
    vector<int> d_path;
    Dijkstra.dijkstra(0, d_path);
    for(int i=0; i<d_path.size(); ++i) cout << i << ": " << d_path[i] << ' ' << endl;

    /* Bellman-Ford */
    cout << "Bellman-Ford:" << endl;
    Graph_bf B_f(5, 8);
    B_f.add_edge(0, 1, -1);
    B_f.add_edge(0, 2, 4);
    B_f.add_edge(1, 2, 3);
    B_f.add_edge(1, 3, 2);
    B_f.add_edge(1, 4, 2);
    B_f.add_edge(3, 2, 5);
    B_f.add_edge(3, 1, 1);
    B_f.add_edge(4, 3, -3);

    // will be same as previous if changed from undirected to directed
    // Graph_bf B_f(9, 14);
    // B_f.add_edge(0, 1, 4);
    // B_f.add_edge(0, 7, 8);
    // B_f.add_edge(1, 2, 8);
    // B_f.add_edge(1, 7, 11);
    // B_f.add_edge(2, 3, 7);
    // B_f.add_edge(2, 5, 4);
    // B_f.add_edge(2, 8, 2);
    // B_f.add_edge(3, 4, 9);
    // B_f.add_edge(3, 5, 14);
    // B_f.add_edge(4, 5, 10);
    // B_f.add_edge(5, 6, 2);
    // B_f.add_edge(6, 7, 1);
    // B_f.add_edge(6, 8, 6);
    // B_f.add_edge(7, 8, 7);

    vector<int> b_path;
    // B_f.print();
    B_f.bell_ford(0, b_path);
    for(int i=0; i<b_path.size(); ++i) cout << i << ": " << b_path[i] << ' ' << endl;

    /* Floyd-Warshall */
    cout << "Floyd-Warshall:" << endl;
    int size_f = 4;
    Graph_fw F_w(size_f);
    F_w.add_edge(0, 1, 5);
    F_w.add_edge(0, 3, 10);
    F_w.add_edge(1, 2, 3);
    F_w.add_edge(2, 3, 1);

    vector<vector<int>> f_path;
    F_w.floyd_warshall(f_path);
    for (int i=0;i<size_f;i++) {
        for (int j=0;j<size_f;j++) {
            if (f_path[i][j] == INF) {
                cout << "INF" << '\t';
            } else {
                cout << f_path[i][j]<< "\t";
            }
        }
        cout << endl;
    }
    /* Kruskal (edged based) */
    cout << "Kruskal (edged based):" << endl;
    Graph_kruskal_n Kruskal(4, 5);
    Kruskal.add_edge(0, 1, 10);
    Kruskal.add_edge(0, 2, 6);
    Kruskal.add_edge(0, 3, 5);
    Kruskal.add_edge(1, 3, 15);
    Kruskal.add_edge(2, 3, 4);

    Kruskal.kruskal();
    Kruskal.print_mst();

    /* Kruskal (matrix based) */
    cout << "Kruskal (matrix based):" << endl;
    Graph_kruskal_m Kruskal_m(false, 4);
    Kruskal_m.add_edge(0, 1, 10);
    Kruskal_m.add_edge(0, 2, 6);
    Kruskal_m.add_edge(0, 3, 5);
    Kruskal_m.add_edge(1, 3, 15);
    Kruskal_m.add_edge(2, 3, 4);

    // Kruskal_m.print();
    Kruskal_m.kruskal();
    Kruskal_m.print_mst();

    /* Prim's Algorithm */
    cout << "Prim:" << endl;
    Graph_prim Prim(false, 4);
    Prim.add_edge(0, 1, 10);
    Prim.add_edge(0, 2, 6);
    Prim.add_edge(0, 3, 5);
    Prim.add_edge(1, 3, 15);
    Prim.add_edge(2, 3, 4);

    // Kruskal_m.print();
    Prim.prim();
    Prim.print_mst();

    return 0;
}
