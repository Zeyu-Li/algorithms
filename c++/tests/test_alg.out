GCD:
21
LCM:
525
Quick Square Root: 
0.315686 : 0.316226
BFS:
Length: 2
BFS order: 0 1 3 2 4 7 5 6
Shortest path: 0 3 7
DFS:
DFS order: 2 0 1 3
Dijkstra:
0: 0
1: 4
2: 12
3: 19
4: 21
5: 11
6: 9
7: 8 
8: 14
Bellman-Ford:
0: 0
1: -1
2: 2
3: -2
4: 1
Floyd-Warshall:
0       5       8       9
INF     0       3       4
INF     INF     0       1
INF     INF     INF     0
Kruskal (edged based):
2 - 3 (4)
0 - 3 (5)
0 - 1 (10)
Kruskal (matrix based):
2 - 3 (4)
0 - 3 (5)
0 - 1 (10)
Prim:
0 - 1 (10)
3 - 2 (4)
0 - 3 (5)
